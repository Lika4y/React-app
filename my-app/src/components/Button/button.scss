@import "../ProductList/myconst.scss";

.button {
    border-radius: 3px;
    border: 1px solid $border-color;
    font-size: $font;
    font-family: $font-family;
    margin: 0 auto;
    font-weight: bold;
    color: $border-color;
    display: flex;
    justify-content: center;

    &:hover {
        cursor: pointer;
        color: black;
    }

    &-wrapper {
        display: flex;
        margin-top: 15px;
    }

}

.delete-btn{
    background-color: rgb(250, 240, 240);
    display: flex;
    align-items: flex-end;
    height: 20px;
    border-radius: 3px;
    border: 1px solid $border-color;
}