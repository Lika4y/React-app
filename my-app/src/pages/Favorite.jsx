import ProductCard from "../components/Productсart/Productcard";
import { useProductList } from "../hooks/useProductList";
import { useDispatch, useSelector } from 'react-redux';
import { removeFavorite } from "../redux/actions/favorites";
import { selectFavorID, selectProducts } from "../redux/selectors";



export function Favorite(){

const dispatch = useDispatch();

    const removeFavor = (e, id) => {
        if(e.target.tagName !== "IMG") return false;
        id = e.target.id;
        dispatch(removeFavorite(id));
    }

let favorId = useSelector(selectFavorID);
let products = useSelector(selectProducts);
let arrfavor = useProductList(favorId, products);


    return(
        <>
        <div className="card-box">
             {!arrfavor.length ? "There are no products in the favorites": arrfavor.map((item) => {
                     return (
                         <div key={item.code} className="card" id={item.code} onClick={removeFavor}>
                                 <ProductCard id={item.code} src={item.url} name={item.name} price={item.price} code={item.code} color={item.color} {...item} />
                         </div>
                     )
                 })}
            </div>
        </>
    ) 
}