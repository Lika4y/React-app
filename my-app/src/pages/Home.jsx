import ProductList from "../components/ProductList/ProductList";
import Modal from "../components/Modal/Modal";
import { addFavorite, removeFavorite } from "../redux/actions/favorites";
import { useDispatch, useSelector } from 'react-redux';
import { closeModal } from "../redux/actions/openModal";
import { addCart } from "../redux/actions/cart";
import { selectCartID, selectCurrent, selectOpen } from "../redux/selectors";


export function Home(props) {
const dispatch = useDispatch();

let open = useSelector(selectOpen);
let current = useSelector(selectCurrent);
let cart = useSelector(selectCartID);

    const addFavorites = (e, id) => {
        if (e.target.tagName !== "IMG") return false;
        id = e.target.id;
        e.target.className === "add-to-favorite" ? dispatch(addFavorite(id)) : dispatch(removeFavorite(id))
    }

    const addedCart = () => { 
        let isInArray = false;
        cart.forEach(item => {
            if(item === current){
                isInArray = true;
                dispatch(closeModal());
            }
        })
        if(!isInArray) {
            dispatch(addCart(current));
        }
        
        dispatch(closeModal());
   }

    return (
        <>
            <ProductList click= {addFavorites} onClick={props.show} />
            <Modal
                className={open ? "modal-wrapper active" : "modal-wrapper"}
                save={addedCart}
                close={() => {dispatch(closeModal())}}
            />
        </>
    )
}