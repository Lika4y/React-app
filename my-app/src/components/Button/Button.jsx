import "./button.scss";

function Button (props){
        return (
            <button data-testid="button-element" id={props.id} data-modal-id={props.code} className="button" style={{backgroundColor: props.color}} onClick = {(e) => {props.onClick(e)}}>{props.text}</button>
        ) 
}

export default Button;

