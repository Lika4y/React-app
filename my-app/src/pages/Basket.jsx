import ProductCard from "../components/Productсart/Productcard";
import Modal from "../components/Modal/Modal";
import "../components/Button/button.scss";
import { useProductList } from "../hooks/useProductList";
import { useDispatch, useSelector } from 'react-redux';
import { openModal, closeModal } from "../redux/actions/openModal";
import { addCurrentProduct } from "../redux/actions/current";
import { removeCart } from "../redux/actions/cart";
import { selectCartID, selectOpen, selectProducts } from "../redux/selectors";
import OrderForm from "../components/Form/Form";
import { addFavorite, removeFavorite } from "../redux/actions/favorites";



export function Basket(props) {
const dispatch = useDispatch();

let open = useSelector(selectOpen);
let products = useSelector(selectProducts);
let cart = useSelector(selectCartID);

let arrcart = useProductList(cart, products);

    const showModalCart = (e) => {
       
        let modalData = e.target.getAttribute("data-num");
        dispatch(addCurrentProduct(modalData));
        dispatch(openModal({
            header: "Removing an Item from the cart", 
            text: "Are you sure you want to remove an item from your shopping cart?", 
            actions: ["Ok", "Cancel"], 
            closeButton: false}));
        }

        let currentproduct = useSelector(state => state.current.currentProduct);


        const removedCart = () => {
                dispatch(removeCart(currentproduct));
                dispatch(closeModal());
                            }

                            const addFavorites = (e, id) => {
                                if (e.target.tagName !== "IMG") return false;
                                id = e.target.id;
                                e.target.className === "add-to-favorite" ? dispatch(addFavorite(id)) : dispatch(removeFavorite(id))
                        
                            }

    return (
        <>
            <div className="card-box" onClick={addFavorites}>
             {!arrcart.length ? "There are no products in the cart" : arrcart.map((item) => {
                     return (
                         <div style={{display:"flex", gap: "5px"}} key={item.code}>
                         <div className="card" >
                                 <ProductCard show={props.show} id={item.code} src={item.url} name={item.name} price={item.price} code={item.code} color={item.color} {...item} />
                                
                                 <Modal className={open ? "modal-wrapper active" : "modal-wrapper"}
                                        save={removedCart}
                                        close={() => {dispatch(closeModal())}}
                                        /> 
                         </div>
                         <button data-testid="basket-button-element" className="delete-btn" data-num={item.code} data-modal-id={2} onClick={showModalCart}>X</button>
                
                         </div>
                     )
                 })}
             </div>
             <div>
                <OrderForm/>
             </div>
        </>
    )

}


