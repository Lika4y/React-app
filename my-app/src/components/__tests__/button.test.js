import React from "react";
import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import Button from "../Button/Button";
import { Basket } from "../../pages/Basket";
import renderer from 'react-test-renderer';
import { useDispatch, useSelector } from "react-redux";
import OrderForm from "../Form/Form";

jest.mock('react-redux');

afterEach(cleanup);

describe("button add to cart", () => {
    test("renders without crashing", () => {
        render(<Button />);
    })

    test("renders add to cart message", () => {
        render(<Button text="Add to cart" />);
        expect(screen.getByText("Add to cart")).toBeInTheDocument();
    })

    test("there is in the document a button", () => {
        render(<Button />);
        expect(screen.getByRole("button")).toBeInTheDocument();
    })

    test('button style color pink', () => {
        const buttonProps = {
            id: 1,
            code: 2,
            color: 'pink',
            text: 'Add to Cart',
            onClick: jest.fn(),
        }
        render(<Button {...buttonProps} />);
        expect(screen.getByTestId('button-element')).toBeInTheDocument();
        expect(screen.getByRole("button")).toHaveStyle("background-color: pink");
    })

    test('snapshot button', () => {
        const buttonProps = {
            id: 1,
            code: 2,
            color: 'pink',
            text: 'Add to Cart',
            onClick: jest.fn(),
        }
        const button = renderer.create(<Button {...buttonProps} />).toJSON();
        expect(button).toMatchSnapshot();
    })

    test("button click", () => {
        const handleChange = jest.fn();
        const { container } = render(<Button onClick={handleChange} />);
        const btn = container.firstChild;
        expect(handleChange).not.toHaveBeenCalled();
        fireEvent.click(btn);
        expect(handleChange).toHaveBeenCalledTimes(1);
    })
})

describe("button in the Basket", () => {
    test("there is in the basket a button delete", () => {
        useSelector.mockReturnValue([]);
        render(<Basket />);
        expect(screen.getByRole("button")).toBeInTheDocument();
    });

    test("button schould click", () => {
        const handleChange = jest.fn();
        useSelector.mockReturnValue([]);
        render(<Basket />);
        let btn = screen.getByRole("button");
        btn.onclick = handleChange;
        expect(handleChange).not.toHaveBeenCalled();
        fireEvent.click(btn);
        expect(handleChange).toHaveBeenCalledTimes(1);
    })

})

describe("button in the Form", () => {
    test("there is in the form a button checkout", () => {
        useSelector.mockReturnValue([]);
        render(<OrderForm />);
        expect(screen.getByRole("button")).toBeInTheDocument();
    })

    test("button schould click", () => {
        const handleChange = jest.fn();
        useSelector.mockReturnValue([]);
        render(<OrderForm />);
        let btn = screen.getByRole("button");
        btn.onclick = handleChange;
        expect(handleChange).not.toHaveBeenCalled();
        fireEvent.click(btn);
        expect(handleChange).toHaveBeenCalledTimes(1);
    })
    
})




