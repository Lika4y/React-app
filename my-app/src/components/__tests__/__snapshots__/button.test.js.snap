// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button add to cart snapshot button 1`] = `
<button
  className="button"
  data-modal-id={2}
  data-testid="button-element"
  id={1}
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "pink",
    }
  }
>
  Add to Cart
</button>
`;
